boot/bootstate20.go

```diff
@@ -0,0 +1,339 @@
+	}
+
+	currentKernel, err := u20.ebl.Kernel()
+	if err != nil {
+		return false, nil, err
+	}
+
+	kernelStatus := "try"
+	rebootRequired = true
+	// check if the next kernel snap is the same as the current kernel snap,
+	// in which case we either don't need to do anything or we just need to
+	// clear the status and not reboot
+	if nextKernel.SnapName() == currentKernel.SnapName() && nextKernel.SnapRevision() == currentKernel.SnapRevision() {
+		// If we were in anything but default ("") mode before
+		// and switched to the good core/kernel again, make
+		// sure to clean the kernel_status here. This also
+		// mitigates https://forum.snapcraft.io/t/5253
+		if u20.env["kernel_status"] == "" {
+			// already clean
+			return false, nil, nil
+		}
+
+		// clean
+		kernelStatus = ""
+		rebootRequired = false
+	} else {
+		// need to enable the symlink
+		u20.snapToTry = nextKernel
+	}
+
+	u20.toCommit["kernel_status"] = kernelStatus
+
+	return rebootRequired, u20, nil
+}
+
+type bootStateUpdate20 struct {
+	typ       snap.Type
+	ebl       bootloader.ExtractedRunKernelImageBootloader
+	env       map[string]string
+	toCommit  map[string]string
+	snapToTry snap.PlaceInfo
+	snapTried snap.PlaceInfo
+}
+
+func newBootStateUpdate20(u bootStateUpdate, typ snap.Type) (*bootStateUpdate20, error) {
+	// TODO:UC20: re-factor to handle base snap update as well as kernel snap
+	// TODO:UC20: handle base_status from modeenv as well
+	if u != nil {
+		u20, ok := u.(*bootStateUpdate20)
+		if !ok {
+			return nil, fmt.Errorf("internal error: threading unexpected boot state update with 20: %T", u)
+		}
+		return u20, nil
+	}
+	bl, err := bootloader.Find("", nil)
+	if err != nil {
+		return nil, err
+	}
+	ebl, ok := bl.(bootloader.ExtractedRunKernelImageBootloader)
+	if !ok {
+		return nil, fmt.Errorf("cannot use %s bootloader: does not support extracted run kernel images", bl.Name())
+	}
+
+	m, err := bl.GetBootVars("kernel_status")
+	if err != nil {
+		return nil, err
+	}
+	return &bootStateUpdate20{
+		ebl:      ebl,
+		env:      m,
+		toCommit: make(map[string]string),
+		typ:      typ,
+	}, nil
+}
+
+func (u20 *bootStateUpdate20) commit() error {
+	// TODO:UC20: handle base_status from modeenv as well
+	if len(u20.toCommit) == 0 {
+		// nothing to do
+		return nil
+	}
+
+	// make a copy of the env map, because we specifically need to check
+	// later on what the original value of kernel_status was that we entered in
+	// with was
+	envCopy := make(map[string]string, len(u20.env))
+	for k, v := range u20.env {
+		envCopy[k] = v
+	}
+	// The ordering of this is very important for boot safety/reliability!!!
+
+	// If we are about to try an update, and need to add the try-kernel symlink,
+	// we need to do things in this order:
+	// 1. Add try-kernel symlink
+	// 2. Update kernel_status to "try"
+	//
+	// This is because if we get rebooted in between 1 and 2, kernel_status is
+	// still "trying", so the bootloader will interpret this to mean the boot
```

Comment:
Isn't the `kernel_status == ""` at this point and the boot script just ignores that and boots with the current kernel?

Reply:
Yes you are correct my comment was wrong